### PR TITLE
[CELEBORN-1356] Split rpc module into rpc_app and rpc_service

### PR DESCRIPTION
**Pull request by mridulm**
### What changes were proposed in this pull request?

Split the `rpc` transport module into `rpc_app` and `rpc_service` to allow for them to be independently configured.

### Why are the changes needed?

We need the ability to independently configure communication between application components (driver/executors in spark applications) and those to/from Celeborn service (master/workers) components.

This is particularly relevant for TLS support where applications might be running with TLS disabled for their rpc services or using self-signed certificates (see CELEBORN-1354 for an example), while services would have signed certs.

### Does this PR introduce _any_ user-facing change?

Yes, it allows users to independently configure rpc env within the application and those to/from services.
Backward compatibility is maintained - and so existing `rpc` is the fallback in case `rpc_app` or `rpc_service` config is not found.

### How was this patch tested?

Unit tests were enhanced, existing tests pass.

### CODE REVIEW

**Comment by mridulm** [comment]
If it makes it easier to review, I can split it into two - updates to `CelebornConf`, which is mostly config updates; and wiring up the new transport module to RpcEnv and usages.
Please let me know if that helps.

**Comment by mridulm** [comment]
The test failure seems to be some timeout issue ? this is ready for review.

+CC @otterc , @waitinfuture, @FMX , @pan3793

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 1]
```diff
@@ -167,6 +231,11 @@ class CelebornConf(loadDefaults: Boolean) extends Cloneable with Logging with Se
     Utils.timeStringAsSeconds(get(key, defaultValue))
   }
 
+  /** Get time in seconds for a transport config */
+  def getTransportConfTimeAsSeconds(module: String, configEntry: ConfigEntry[Long]): Long = {
```
Comment (FMX) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:235]: This method is not used.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 2]
```diff
@@ -251,6 +335,11 @@ class CelebornConf(loadDefaults: Boolean) extends Cloneable with Logging with Se
     Utils.byteStringAsMb(get(key, defaultValue))
   }
 
+  /** Get a size parameter as mb for a transport config */
+  def getTransportConfSizeAsMb(module: String, configEntry: ConfigEntry[Long]): Long = {
```
Comment (FMX) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:339]: This method is not used too, I think adding a new method in time will be better.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 3]
```diff
@@ -309,6 +408,11 @@ class CelebornConf(loadDefaults: Boolean) extends Cloneable with Logging with Se
     getOption(key).map(_.toLong).getOrElse(defaultValue)
   }
 
+  /** Get a parameter as a long for a transport config */
+  def getTransportConfLong(module: String, configEntry: ConfigEntry[Long]): Long = {
```
Comment (FMX) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:412]: ditto.

**Review by FMX — COMMENTED** [review]
I think split rpc module looks fine to me, but there is no need to introduce something not used currently.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 1 above)
Comment (mridulm) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:235]: Agree, I added those for completeness sake - but I can remove all those which are not used.
Will update the PR

Comment (mridulm) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala]: Done

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 2 above)
Comment (mridulm) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala]: Done

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 3 above)
Comment (mridulm) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala]: Done

File: common/src/main/java/org/apache/celeborn/common/protocol/TransportModuleConstants.java [hunk 4]
```diff
@@ -21,6 +21,17 @@ public class TransportModuleConstants {
...
   public static final String REPLICATE_MODULE = "replicate";
   public static final String FETCH_MODULE = "fetch";
-  public static final String RPC_MODULE = "rpc";
+
+  // RPC module used by the application components to communicate with each other
+  // This is used only at the application side.
+  public static final String RPC_APP_MODULE = "rpc_app";
+  // RPC module used to communicate with/between server components
+  // This is used both at server (master/worker) and application side.
+  public static final String RPC_SERVER_MODULE = "rpc_server";
```
Comment (pan3793) [inline common/src/main/java/org/apache/celeborn/common/protocol/TransportModuleConstants.java:30]: maybe "app:service" and "client:server", we used the term pair "client:server" in some places, for example, the master client could be either on the app or worker.

Comment (mridulm) [reply common/src/main/java/org/apache/celeborn/common/protocol/TransportModuleConstants.java:30]: Doesn't this not need to indicate it is specifically for rpc ? and not for other transport modules ?

Comment (pan3793) [reply common/src/main/java/org/apache/celeborn/common/protocol/TransportModuleConstants.java:30]: sorry for misleading, I mean rpc_app and rpc_service

Comment (mridulm) [reply common/src/main/java/org/apache/celeborn/common/protocol/TransportModuleConstants.java:30]: Ah, service makes sense - agree.
Will rename

Comment (mridulm) [reply common/src/main/java/org/apache/celeborn/common/protocol/TransportModuleConstants.java]: Done, PTAL

**Comment by pan3793** [comment]
code change lgtm, could you pls add an item to the migration guide?

**Review by pan3793 — APPROVED** [review]

**Comment by mridulm** [comment]
> code change lgtm, could you pls add an item to the migration guide?

I was planning to add that as part of https://issues.apache.org/jira/browse/CELEBORN-1353 .. should I be adding it as part of this PR instead ?

**Comment by pan3793** [comment]
@mridulm thanks for opening a dedicated ticket for docs, merged to main for 0.5

**Comment by mridulm** [comment]
Thanks for the reviews @FMX and @pan3793 !

**Pull request closed without merging**